compiler/rustc_mir/src/transform/instrument_coverage.rs

```diff
@@ -787,7 +793,11 @@ fn filtered_statement_span(statement: &'a Statement<'tcx>, body_span: Span) -> O
         // These statements have spans that are often outside the scope of the executed source code
```

Comment:
@tmandry @wesleywiser - I don't think this is practical at the current time. There are very few special cases now. (There used to be more, but I was able to remove most of them.) The special cases that are left are (for the most part) necessary right now.

I think the only way to avoid this in the long term would require removing the spans from `StorageLive` and `StorageDead` (which I doubt is feasible). But that's just moving the special case handling outside of the InstrumentCoverage pass.

From a coverage perspective, these statements are problematic because their spans refer to the variable declarations, which are quite often in a completely different area in the source from where they actually need to come alive or die.

(Control Flow vs. Data Flow)

As for the other "special cases", I could remove `Coverage` and `Nop`. They are just included for completeness, but should not crop up here anyway.

The `ForGuardBinding` special case is definitely something I would like to remove, but there's already a FIXME for that. That's there because there is a bug somewhere else in the MIR code (I guess) producing an invalid Span.

Reply:
(I'm actually amazed that these are the only special cases. It's way better than I expected, and I'm impressed that the MIR generally stays pretty true to the original source code.)